### PR TITLE
nautilus: os/bluestore: implement Hybrid allocator

### PR DESCRIPTION
**Pull request by ifed01**
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/test/objectstore/Allocator_bench.cc [hunk 1]
```diff
@@ -329,10 +329,42 @@ TEST_P(AllocTest, test_alloc_bench_10_300)
...
+  delete(alloc);
+  ASSERT_EQ(mempool::bluestore_alloc::allocated_bytes(), bytes);
+  ASSERT_EQ(mempool::bluestore_alloc::allocated_items(), items);
+}
+
 INSTANTIATE_TEST_CASE_P(
   Allocator,
   AllocTest,
-  ::testing::Values("stupid", "bitmap"));
+  ::testing::Values("stupid", "bitmap, avl", "hybrid"));
```
Comment (aclamk) [inline src/test/objectstore/Allocator_bench.cc:367]: typo!

File: src/common/options.cc [hunk 2]
```diff
@@ -4949,15 +4950,23 @@ std::vector<Option> get_global_options() {
     .set_description("Determines bluefs volume selection policy")
     .set_long_description("Determines bluefs volume selection policy. 'use_some_extra' policy allows to override RocksDB level granularity and put high level's data to faster device even when the level doesn't completely fit there"),
 
-    Option("bluestore_volume_selection_reserved_factor", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
-      .set_flag(Option::FLAG_STARTUP)
-      .set_default(2.0)
-      .set_description("DB level size multiplier. Determines amount of space at DB device to bar from the usage when 'use some extra' policy is in action. Reserved size is determined as sum(L_max_size[0], L_max_size[L-1]) + L_max_size[L] * this_factor"),
+    Option("bluestore_avl_alloc_bf_threshold", Option::TYPE_UINT, Option::LEVEL_DEV)
+    .set_default(131072)
+    .set_description(""),
```
Comment (aclamk) [inline src/common/options.cc:4955]: We should have a description.

File: src/common/options.cc [hunk 3]
```diff
@@ -4949,15 +4950,23 @@ std::vector<Option> get_global_options() {
...
+
+    Option("bluestore_avl_alloc_bf_free_pct", Option::TYPE_UINT, Option::LEVEL_DEV)
+    .set_default(4)
+    .set_description(""),
```
Comment (aclamk) [inline src/common/options.cc:4959]: We should have a description.

File: src/common/options.cc (hunk 2 above)
Comment (ifed01) [reply src/common/options.cc:4965]: This is a backport PR so it doesn't bring additional stuff in. We should have another PR against master for that first.

File: src/common/options.cc (hunk 3 above)
Comment (ifed01) [reply src/common/options.cc:4969]: same here

Comment (aclamk) [reply src/common/options.cc:4969]: @ifed01 True, I will make PR with description in this.

**Comment by ifed01** [comment]
jenkins test make check

**Comment by neha-ojha** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
jenkins test dashboard backend

**Comment by smithfarm** [comment]
@ifed01 @aclamk @neha-ojha @jdurgin Putting this in the queue for integration testing and merge for the upcoming 14.2.11 release.

**Comment by yuriw** [comment]
[wip-yuri-testing-2020-07-17-1818-nautilus](https://trello.com/c/YeaMdnHD/967-wip-yuri-testing-2020-07-17-1818-nautilus)

**Comment by yuriw** [comment]
@ifed01 this passed the test and was approved by @neha-ojha 
pls resolve conflict and merge at will

**Pull request merged**